### PR TITLE
Generalize `scan_prefix` to scan a subrange within the prefix

### LINKED ISSUE #1778: Prefix bloom filters are not used for normal range scans
SlateDB supports prefix bloom filters, and `scan_prefix` can use them to skip SSTs that do not contain the requested prefix. However, normal bounded range scans do not currently leverage prefix bloom filters, even when the range clearly targets a prefix-shaped keyspace.

For example, a scan like:

```rust
db.scan(b"user:123:"..b"user:124:")
```

may be semantically similar to a prefix scan over `user:123:`, but today it goes through the normal range-scan path and does not pass a prefix query to SST filter evaluation. This can make range scans over prefix-partitioned keys scale with the number of overlapping SSTs even when configured prefix bloom filters could rule many of them out.

This is related to the broader scan latency problem discussed in #1302 and the prefix filtering work in #1334, but the remaining gap is specifically that `scan` / `scan_with_options` cannot benefit from prefix bloom filters.

Possible directions:
- Add an explicit scan option that lets callers provide a prefix-filter hint for a range scan, with validation that the requested range is contained within that prefix range.
- Explore safe inference for ranges whose bounds unambiguously describe a single prefix interval, while preserving the current non-inference behavior where ambiguous.

The goal would be to let users express prefix-shaped bounded scans without losing the SST pruning benefits of prefix bloom filters.

### PR DESCRIPTION
**Pull request by apurvam**
Fixes #1778.

## Summary

`scan_prefix` and `scan_prefix_with_options` take a second argument that restricts  the scan to a range of key suffixes within the prefix. A bound `s` selects the  full key `prefix ++ s`, and `..` keeps the old behavior of scanning every key with the prefix.

```rust
  db.scan_prefix(b"user1:", ..).await?;                // every key under user1:
  db.scan_prefix(b"user1:", b"0005"..).await?;         // user1:0005 onward
  db.scan_prefix(b"user1:", b"0005"..b"0042").await?;  // a bounded slice
```

This is a breaking change: existing callers become `scan_prefix(prefix, ..)`.

##  Why

Consider keys laid out as `prefix ++ sequence_number`, with a reader that wants everything under one prefix from sequence N onward. Before this change there were two ways to express that:

* `scan(prefix_plus_start..prefix_plus_end)`, built by hand-concatenating the prefix onto each bound, reads exactly the right byte range but never consults prefix bloom filters. Any SST whose key range overlaps the scan gets opened, including SSTs that hold no keys with the prefix at all. 
* `scan_prefix(prefix)` consults the filters and skips those SSTs, but it scans  the prefix's whole keyspace from the beginning, and the caller discards  everything before N.

The new call combines both. The filter check skips whole SSTs before any index or data block is read, and the narrowed range positions block reads at the subrange start within the SSTs that remain. Consulting the filter is sound here because the composed range is built from the prefix, so every key in it carries the prefix by construction. A plain scan cannot guarantee that about its range, which is why it does not consult filters.

##  The subrange parameter type

The subrange is `impl SubrangeBounds`, a small new trait, rather than the generic `RangeBounds<K>` that scan takes. Neither standard option works here:

* With generic `RangeBounds<K>`, the most common call, `scan_prefix(prefix, ..)`, does not compile: `..` carries no bound values, so the compiler cannot determine  K without a written-out type annotation. The same problem is why every  full-range scan(..) in this codebase carries a `::<Vec<u8>, _>` annotation.
* With concrete `RangeBounds<&[u8]>`, `..` compiles but natural arguments do not: the bounds of `b"0005"..b"0042"` are `&[u8; 4]` rather than `&[u8]`, and ranges over `Vec<u8>` or `Bytes` are rejected too. Every bound would need hand conversion, as in `b"0005".as_slice()..b"0042".as_slice()`.

`SubrangeBounds` mirrors `RangeBounds` but always yields its bounds as byte slices. `..` has a dedicated implementation, so it needs no annotation, and the other standard range types accept any AsRef<[u8]> bound. Byte literals, slices, `Vec<u8>`, `Bytes`, and explicit Bound pairs all work as written.

`Db`, `DbReader`, `DbSnapshot`, `DbTransaction`, and the `DbReadOps` trait each funnel plain scans and prefix scans through one shared inner path, and the only difference between them is whether a prefix is present. Under serializable-snapshot isolation,  transactions track the composed range: a concurrent write inside the subrange aborts the transaction, while a write under the same prefix but outside the  subrange commits cleanly.

##  Additional bugfix

Range scans built their L0 iterators without DbStats, so prefix-filter skips on L0 SSTs happened but were never counted, while the sorted-run iterators counted theirs. The new integration test exposed this by showing that the counter of filter-skipped SSTs stayed 0 even though SSTs were being skipped. Fixed by threading the stats handle through `build_l0_range_iters`.

##  Benchmarks

 The numbers previously reported (80% fewer object store GETs at high segment cardinality, 73x fewer bytes read on tail scans) were measured under a version where the prefix was inferred from the Range bounds. For that workload the composed range is identical to the range with this explicit API so the results should carry over. I'll have to update the benchmark to rerun it, which I can do in parallel with this review.

## Testing

* Unit tests in `bytes_range` cover every combination of subrange bound kinds (inclusive, exclusive, and unbounded on each end) along with the empty-prefix and all-0xFF-prefix edge cases. Property tests check that `from_prefix_and_subrange(p, ..)` equals `from_prefix(p)` and that every key sampled from a composed range starts with the prefix. That last property is what makes it safe to act on a filter miss: an SST whose filter has no entry for the prefix cannot contain any key in the range.
* An integration test arranges two SSTs whose key ranges overlap a bbb scan but which hold no bbb keys. Range pruning cannot exclude them, so the only way they get skipped is the filter; the test asserts the skip counter rises by exactly two. The same test asserts a plain scan over identical bounds does not.
* An existing test helper runs one set of scan assertions against each read handle (`Db`, `DbSnapshot`, and `DbReader`, called through the `DbReadOps` trait). It now includes bounded-subrange cases, so each handle is checked to pass the subrange through correctly rather than only the full-prefix .. case.
* A transaction test checks exactly which range gets recorded for conflict detection: a write inside the subrange aborts the transaction even though the scan never returned that key, and a write under the same prefix but outside the subrange does not.
* Property tests run the same subrange scans on two databases, one built with prefix filters and one without, and also compare both against filtering a full prefix scan client-side.

### CODE REVIEW

File: slatedb/src/bytes_range.rs [hunk 1]
```diff
@@ -117,6 +117,56 @@ impl BytesRange {
...
+    ///   "increment" of a longer prefix of the start bound (e.g.
+    ///   `b"user1:42".."user2"` infers `user1`, not just `user`): the range
+    ///   then sits inside that prefix's half-open keyspace even though the
+    ///   bounds' common prefix is shorter. This is the natural shape of
+    ///   "scan from a cursor to the end of a logical prefix".
+    ///
+    /// The inferred prefix is a conservative lower bound on specificity, not
+    /// necessarily the longest theoretically valid one; configured prefix
+    /// extractors truncate it to their canonical length anyway.
+    pub(crate) fn infer_prefix(&self) -> Option<Bytes> {
```
Comment (hachikuji) [inline slatedb/src/bytes_range.rs]: I wonder if it is better to lean on the prefix extractor to tell us how to build prefixes from range bounds. Say that I had a key structure such as `<a><b><c>`. I might choose `<a>` as the prefix for the filter, but it may still be useful to query based on `<a><b>`. Longest prefix inference would leave us with `<a><b>` and the filter would be wasted. The prefix extractor instead could tell us explicitly to query within `<a>` and we don't have to guess.

On the other hand, the API to extract the prefix from a general range may be awkward because you need both bounds. I'm starting to think that a cleaner api might be to extend `scan_prefix` to support a sub-range or something. Something like `scan_prefix_range(prefix, subrange)`.

Comment (criccomini) [reply slatedb/src/bytes_range.rs]: cc @nomiero

Comment (hachikuji) [reply slatedb/src/bytes_range.rs]: My example was not great because the current `scan_prefix` can already handle `<a><b>` queries. The gap is when you need to query within a subset of a prefix. Say for example, scanning `<c:500>..<c:1000>` within an `<a><b>` prefix. I think the main point still stands though. Relying on inference would lead to the wrong prefix getting extracted. With something like `scan_prefix_range`, the user can specify the query as `scan_prefix_range(<a:1><b:2>, <c:500>..<c:1000>)`. Then we can extract the `<a>` prefix correctly and the interpretation is unambiguous for the extractor.

File: slatedb/src/db_transaction.rs [hunk 2]
```diff
@@ -226,9 +226,11 @@ impl DbTransaction {
...
             .end_bound()
             .map(|b| Bytes::copy_from_slice(b.as_ref()));
-        let range = (start, end);
-        self.scan_inner(BytesRange::from(range), options, None)
-            .await
+        let range = BytesRange::from((start, end));
+        // Consult prefix bloom filters when a prefix is inferable from the
+        // range bounds; see `Db::scan_with_options`.
+        let prefix = range.infer_prefix();
+        self.scan_inner(range, options, prefix).await
```
Comment (criccomini) [inline slatedb/src/db_transaction.rs]: Does it make sense to push the infer_prefix call into `scan_inner` so all invocations use it?

Same question for the self.inner.scan_with_options call above.

File: slatedb/src/bytes_range.rs (hunk 1 above)
Comment (criccomini) [reply slatedb/src/bytes_range.rs]: @hachikuji I kinda agree. I also think it'd be a cleaner API. It would, for example, make it easier for us to skip all this logic if no extractor is configured. In general, I prefer using the extractor explicitly as much as possible vs. trying to infer things. I worry a bit that the baked in assumptions will lead to bugs later.

Comment (nomiero) [reply slatedb/src/bytes_range.rs]: I'm a bit late to the discussion, but I agree with the comments here. We had a discussion on this in the RFC for prefix bloom filter and reached a similar conclusion as well. An API like `scan_prefix_range(prefix, subrange)` is better. It can benefit from range pruning other than the prefix bloom filter and it's a cleaner explicit contract.

Comment (apurvam) [reply slatedb/src/bytes_range.rs]: Sounds good. I'll incorporate this suggestion and add a `scan_prefix_range` api.

**Comment by apurvam** [comment]
@nomiero @criccomini @hachikuji 

wanted to mention @nomiero's suggestion on discord here: 

> 
> For the sake of covering the options for the API for a range for prefix scan
> Which one seems like the best API here. 
> A new API with  scan_prefix_range
> Having scan_prefix and scan_prefix_with_options take an Option for the range.
> Having only scan_prefix_with_options take ScanPrefixOptions instead of ScanOptions and the ScanPrefixOptions would have an optional range.
> 
> These all will work. I'm leaning a bit towards option 2 now. WYT ?
> Only thing I have with this option is it will make scan_prefix normal case be scan_prefix(prefix, None). Not amazing but seems ok. 
> 
> Option 3 is valid, but it will be a bit hidden. 
> 
> Not a big deal, but just wanted to see if this is worth a discussion

Do you prefer `scan_prefix_with_options` or an explicit `scan_prefix_range`. 

If we do `scan_prefix_with_options` then we would have to change all the call sites and docs of `scan_prefix` to include the `None` option argument. 

Further, for `scan_prefix(prefix, None)`, the warning I got from Fable is: 

> scan_prefix(prefix, None) isn't just ugly — None for an unconstrained T: RangeBounds<K> won't type-infer; callers would have to write None::<Range<&[u8]>>. The fix is a bare impl RangeBounds<K> parameter so the common case reads scan_prefix(prefix, ..).

My rust isn't good enough to judge that warning, so I'd love your advice.

**Comment by hachikuji** [comment]
Generalizing `scan_prefix` seems reasonable to me to include the subrange. Using `scan_prefix(prefix, ..)` feels ergonomic to me as a substitute for the current semantic. I'm also ok with `scan_prefix_range` as well. Just annoying if we end up needing `**_with_options` variants for both of them. I do kind of dislike hiding the range in the options for what it's worth.

**Comment by nomiero** [comment]
> scan_prefix(prefix, None) isn't just ugly — None for an unconstrained T: RangeBounds won't type-infer; callers would have to write None::<Range<&[u8]>>

> Generalizing scan_prefix seems reasonable to me to include the subrange. Using scan_prefix(prefix, ..) feels ergonomic to me as a substitute for the current semantic

Yeah, I guess None doesn't work here because a type is needed. using `scan_prefix(prefix, ..)` looks good to me.

**Comment by apurvam** [comment]
@nomiero @hachikuji thanks for chiming in. I'll introduce `scan_prefix_with_options` and introduce `scan_prefix(prefix ..)` for the vanilla scan_prefix. In this world, the range will be in the Options argument.

**Comment by apurvam** [comment]
> In this world, the range will be in the Options argument.

Or would it be better to have a explicit subrange in the `_with_options` variant: `scan_prefix_with_options(prefix, subrange,  &options)`?

**Comment by nomiero** [comment]
Sorry, I think my comment wasn't clear. 

> using scan_prefix(prefix, ..) looks good to me.

By that I meant the existing `scan_prefix` and `scan_prefix_with_options` will have the first two parameters as `(prefix, subrange)` and the breaking change here is that users already using `scan_prefix` will pass the full range `..`. So no new APIs are introduced here.

**Comment by apurvam** [comment]
Ah. That makes sense. thanks for the clarification @nomiero

**Comment by criccomini** [comment]
> Sorry, I think my comment wasn't clear.
> 
> > using scan_prefix(prefix, ..) looks good to me.
> 
> By that I meant the existing `scan_prefix` and `scan_prefix_with_options` will have the first two parameters as `(prefix, subrange)` and the breaking change here is that users already using `scan_prefix` will pass the full range `..`. So no new APIs are introduced here.

SGTM!

**Comment by apurvam** [comment]
I've pushed the new commits implementing what we agreed on above. A summary of the changes:

The API is as @nomiero described. `scan_prefix` and `scan_prefix_with_options` take `(prefix, subrange, ...)`, where the subrange bounds are key suffixes relative to the prefix, and existing callers pass `..` to keep their current behavior. So filters are only consulted with a prefix the caller named, and plain scans carry no prefix logic at all.

Re @criccomini's `scan_inner` comment: the `infer_prefix` half is moot now that inference is gone, but I did consolidate the paths. Each handle routes plain and prefix scans through one shared inner that takes `prefix: Option<Bytes>`.

One addition beyond what we discussed: the subrange parameter is a small new trait, `SubrangeBounds`, instead of `RangeBounds`. A generic `RangeBounds<K>` breaks the most common call, `scan_prefix(prefix, ..)`, because
  `..` gives the compiler nothing to determine `K` from and a concrete `RangeBounds<&[u8]>` rejects ordinary arguments like `b"a"..b"b"` or ranges over `Vec<u8>`, forcing `.as_slice()` on every bound. The trait mirrors `RangeBounds` but always yields byte-slice bounds, so `..`, byte literals, `Vec<u8>`, `Bytes`, and explicit `Bound` pairs all work as written.

Also included is an additional bugfix. Range scans built their L0 iterators without the stats handle, so filter skips on L0 SSTs happened but were never counted, while the sorted-run iterators counted theirs. One of the new  integration tests caught it.

File: slatedb/src/db.rs [hunk 3]
```diff
@@ -1099,29 +1091,44 @@ impl Db {
...
     /// ```
-    pub async fn scan_prefix<P>(&self, prefix: P) -> Result<DbIterator, crate::Error>
+    pub async fn scan_prefix<P, T>(
+        &self,
+        prefix: P,
+        subrange: T,
+    ) -> Result<DbIterator, crate::Error>
     where
         P: AsRef<[u8]> + Send,
+        T: SubrangeBounds + Send,
```
Comment (hachikuji) [inline slatedb/src/db.rs]: Do we need a dedicated type for subranges? It feels like we should try to use the same type that would be used for ranges in `scan`. If the concrete trait makes for more ergonomic usage, could we generalize it (say `ByteRangeBounds`) and use it in `scan`?

File: bindings/uniffi/src/db.rs [hunk 4]
```diff
@@ -107,7 +107,7 @@ impl Db {
 
     /// Scans rows whose keys start with `prefix`.
     pub async fn scan_prefix(&self, prefix: Vec<u8>) -> Result<Arc<DbIterator>, Error> {
-        let iter = self.inner.scan_prefix(prefix).await?;
+        let iter = self.inner.scan_prefix(prefix, ..).await?;
```
Comment (hachikuji) [inline bindings/uniffi/src/db.rs:110]: Perhaps in a follow-up, we could expose the range here as well?

**Review by hachikuji — COMMENTED** [review]
The patch looks good. I just had one question about the subrange bound type.

File: slatedb/src/db.rs (hunk 3 above)
Comment (apurvam) [reply slatedb/src/db.rs]: Ah. I think that the regular scan suffers from the same issues around ergonomics then. I think we should definitely unify the too. Would it make sense to revert the `SubrangeBounds` in this PR, and then have a followup introducing `ByteRangeBounds` and then using it n `scan` and `scan_prefix`?

That way this PR is focused on just the `scan_prefix` optimization.

File: bindings/uniffi/src/db.rs (hunk 4 above)
Comment (apurvam) [reply bindings/uniffi/src/db.rs:110]: Yes, makes sense. Especially if we introduce `ByteRangeBounds` in a follow up. then we can change the uniffi bindings once.

File: slatedb/src/db.rs (hunk 3 above)
Comment (hachikuji) [reply slatedb/src/db.rs]: Yes, sounds good to me.

Comment (apurvam) [reply slatedb/src/db.rs]: I backed out `SubRangeBounds` and filed #1800 to track the `ByteRangeBounds` suggestion.

File: bindings/uniffi/src/db.rs (hunk 4 above)
Comment (apurvam) [reply bindings/uniffi/src/db.rs:110]: Filed #1801 to track.

File: slatedb/src/db_snapshot.rs [hunk 5]
```diff
@@ -148,14 +147,14 @@ impl DbSnapshot {
...
-    pub async fn scan_prefix<P, T>(
+    pub async fn scan_prefix<'a, P, T>(
         &self,
         prefix: P,
         subrange: T,
     ) -> Result<DbIterator, crate::Error>
     where
         P: AsRef<[u8]> + Send,
-        T: SubrangeBounds + Send,
+        T: RangeBounds<&'a [u8]> + Send,
```
Comment (hachikuji) [inline slatedb/src/db_snapshot.rs:157]: It might not matter too much if we are going to use the new trait, but an alternative here is to use `RangeBounds<P>`, or to introduce a separate `K: AsRef<[u8]> + Send` type if we don't want to force the prefix and subrange to use the same type. The generic-typed `AsRef<[u8]>` is what we currently use in `scan`.

Comment (apurvam) [reply slatedb/src/db_snapshot.rs:157]: I had claude check, and it came back with the following: 

What the three options actually do (rustc-confirmed):

For the two calls that matter — the common full scan scan_prefix(prefix, ..) and a bounded subrange with plain literals scan_prefix(b"ab", b"x"..b"y"):


| subrange bound | `scan_prefix(prefix, ..)` | `scan_prefix(b"ab",  b"x"..b"y")` |
|---|---|---|
| **current** `RangeBounds<&[u8]>` | ✅ compiles | ❌ needs `.as_slice()` on each bound |
| **opt 1** `RangeBounds<P>` | ✅ compiles clean | ❌ `Range<&[u8;1]>: RangeBounds<&[u8;2]>` not satisfied |
| **opt 2** generic `K` (like `scan`) | ❌ **E0283**, needs turbofish | ✅ compiles |

* End claude text

Given this, I think it may make sense to just wait for the proper fix.

Comment (hachikuji) [reply slatedb/src/db_snapshot.rs:157]: Fair enough. I was thinking about consistency with `scan` more so than ergonomics. Today, `scan(..)` doesn't work without a `Vec<u8>` cast. The advantage is that `AsRef` is more general and we can drop the lifetime. However, if we take `ByteRangeBounds`, then both apis will be consistent again.

**Review by hachikuji — APPROVED** [review]
LGTM!
